src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java

```diff
@@ -0,0 +1,340 @@
+        FlowExecution fe = run.getExecution();
+        AtomicBoolean hasExecutionInList = new AtomicBoolean(false);
+        FlowExecutionList.get().forEach(f -> {
+            if (fe != null && f == fe) {
+                hasExecutionInList.set(true);
+            }
+        });
+        if (!hasExecutionInList.get()) {
+            Assert.fail("Build completed but should still show in FlowExecutionList");
+        }
+        CpsFlowExecution cpsExec = (CpsFlowExecution)fe;
+        Assert.assertFalse(cpsExec.isComplete());
+        Assert.assertEquals(Boolean.FALSE, cpsExec.done);
+        Assert.assertFalse(cpsExec.getCurrentHeads().get(0) instanceof FlowEndNode);
+        Assert.assertTrue(cpsExec.startNodes != null && !cpsExec.startNodes.isEmpty());
+    }
+
+    static void assertResultMatchExecutionAndRun(WorkflowRun run, Result[] executionAndBuildResult) throws Exception {
+        Assert.assertEquals(executionAndBuildResult[0], ((CpsFlowExecution)(run.getExecution())).getResult());
+        Assert.assertEquals(executionAndBuildResult[1], run.getResult());
+    }
+
+    /** Create and run a basic build before we mangle its persisted contents.  Stores job number to jobIdNumber index 0. */
+    private static WorkflowRun runBasicBuild(JenkinsRule j, String jobName, int[] jobIdNumber, FlowDurabilityHint hint) throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, jobName);
+        job.setDefinition(new CpsFlowDefinition("echo 'doSomething'", true));
+        job.addProperty(new DurabilityHintJobProperty(hint));
+        WorkflowRun run = j.buildAndAssertSuccess(job);
+        jobIdNumber[0] = run.getNumber();
+        assertCompletedCleanly(run);
+        return run;
+    }
+
+    /** Create and run a basic build before we mangle its persisted contents.  Stores job number to jobIdNumber index 0. */
+    private static WorkflowRun runBasicBuild(JenkinsRule j, String jobName, int[] jobIdNumber) throws Exception {
+        return runBasicBuild(j, jobName, jobIdNumber, FlowDurabilityHint.MAX_SURVIVABILITY);
+    }
+
+    /** Sets up a running build that is waiting on input. */
+    private static WorkflowRun runBasicPauseOnInput(JenkinsRule j, String jobName, int[] jobIdNumber, FlowDurabilityHint durabilityHint) throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, jobName);
+        job.setDefinition(new CpsFlowDefinition("input 'pause'", true));
+        job.addProperty(new DurabilityHintJobProperty(durabilityHint));
+
+        WorkflowRun run = job.scheduleBuild2(0).getStartCondition().get();
+        ListenableFuture<FlowExecution> listener = run.getExecutionPromise();
+        FlowExecution exec = listener.get();
+        while(exec.getCurrentHeads().isEmpty() || (exec.getCurrentHeads().get(0) instanceof FlowStartNode)) {  // Wait until input step starts
+            System.out.println("Waiting for input step to begin");
+            Thread.sleep(50);
```

Comment:
Hmm. `SemaphoreStep` is generally preferable for these kinds of tests. Is there a good reason to use `input` instead?

Reply:
@jglick Yes, Semaphore steps block the CPS thread - input steps follow a more normal pattern.  The durability tests use sleep + semaphore for similar reasons (though sleep should be replaced with input to reduce test times in the near future).

Reply:
No, `SemaphoreStep` does not block the CPS thread. Semantically it is very similar to `input`, but much easier to manipulate from test code.

Reply:
@jglick It definitely blocks *something* and the GraphListeners don't evaluate as they ought to.